src/python/pants/backend/project_info/depgraph.py

```diff
@@ -0,0 +1,187 @@
+    of all possible targets and dependencies in the repo.
+
+    A vertex is identified by an opaque id. These ids are not guaranteed to be stable
+    across multiple runs of the graph creation code.  There are useful member functions
+    that query the graph in terms of Vertex objects alone, so that callers don't have to
+    directly handle ids.
+
+    An edge V1->V2 indicates that V1 depends on V2. An edge may have data attached to it.
+    """
+
+    vertices: tuple[Vertex, ...]
+    edges: tuple[Edge, ...]
+
+    # Redundant data structures useful for querying the graph.
+    vertices_by_id: FrozenDict[str, Vertex]
+    edges_by_src: FrozenDict[str, tuple[Edge, ...]]
+    edges_by_dep: FrozenDict[str, tuple[Edge, ...]]
+
+    def __init__(self, vertices: Iterable[Vertex], edges: Iterable[Edge]):
+        self.vertices = tuple(vertices)
+        self.edges = tuple(edges)
+        self.vertices_by_id = FrozenDict({v.id: v for v in self.vertices})
+        edges_by_src = defaultdict(list)
+        edges_by_dep = defaultdict(list)
+        for edge in self.edges:
+            edges_by_src[edge.src_id].append(edge)
+            edges_by_dep[edge.dep_id].append(edge)
+        self.edges_by_src = FrozenDict({k: tuple(v) for k, v in edges_by_src.items()})
+        self.edges_by_dep = FrozenDict({k: tuple(v) for k, v in edges_by_dep.items()})
+
+    def get_dependencies(self, v: Vertex) -> tuple[Vertex, ...]:
+        return tuple(self.vertices_by_id[edge.dep_id] for edge in self.edges_by_src.get(v.id, []))
+
+    def get_dependees(self, v: Vertex) -> tuple[Vertex, ...]:
+        return tuple(self.vertices_by_id[edge.src_id] for edge in self.edges_by_dep.get(v.id, []))
+
+    def to_json(self) -> str:
+        return json.dumps(self, cls=DependencyGraphJSONEncoder, indent=2)
+
+
+class DependencyGraphJSONEncoder(JSONEncoder):
+    def default(self, o):
+        if isinstance(o, DependencyGraph):
+            return {"vertices": o.vertices, "edges": o.edges}
+        elif isinstance(o, Vertex):
+            return {"id": o.id, "data": o.data}
+        elif isinstance(o, Edge):
+            return {"src_id": o.src_id, "dep_id": o.dep_id, "data": o.data}
+        elif isinstance(o, FrozenDict):
+            return dict(o)  # Unfortunately, FrozenDict is not JSON serializable.
+        return super().default(o)
+
+
+@dataclass(frozen=True)
+class DependencyGraphRequest:
+    addresses: Addresses
+    transitive: bool
+
+
+@rule
+async def generate_graph(request: DependencyGraphRequest) -> DependencyGraph:
+    def target_sorter(tgts_to_sort: Iterable[Target]) -> list[Target]:
+        return sorted(tgts_to_sort, key=lambda t: t.address)
+
+    if request.transitive:
+        transitive_targets = await Get(
+            TransitiveTargets,
+            TransitiveTargetsRequest(request.addresses, include_special_cased_deps=True),
+        )
+        targets = set(transitive_targets.roots) | set(transitive_targets.dependencies)
```

Comment:
Can use `transitive_targets.closure` for this.

Reply:
D'oh! Of course.